internal/cache/resource_map.go

```diff
@@ -1,153 +1,126 @@
-		e.lock.RLock()
-		if !e.isDeleted {
-			// The entry was already present in the map, execute the compute function safely since the read lock is
-			// currently held.
-			compute(key, e.value)
-			// Nothing left to be done, exit the function
-			e.lock.RUnlock()
+		compute(v)
+	})
+
+	return v
+}
+
+// ComputeIfPresent invokes the given function only if a corresponding entry exists in the map for
+// the given key.
+func (m *ResourceMap[K, V]) ComputeIfPresent(key K, compute func(value V)) {
+	m.withLock(key, func() {
+		v, ok := m.load(key)
+		if !ok {
 			return
 		}
 
-		// Very unhappy path: the entry was deleted between this invocation reading it from the map and
-		// acquiring the read lock! Attempt to insert the new entry again by continuing the loop. Note that this
-		// condition is the reason this is in a loop in the first place. It should also be noted that it is
-		// extremely unlikely for this occur, and in most instances this branch will never be reached.
-		e.lock.RUnlock()
-	}
-
-	// Initialize the value while holding the write lock, otherwise a call to ComputeIfPresent could read e.value as
-	// nil.
-	e.value = newValue(key)
-	// Technically, at this point, the write lock does not need to be held anymore, only the read lock since e.value
-	// is initialized. It is however impossible to downgrade the lock write to read without actually releasing it. In
-	// between releasing the write lock and acquiring the read lock, it is possible for the entry to be deleted! Hence,
-	// the write lock is kept.
-	compute(key, e.value)
+		compute(v)
+	})
 }
 
-// DeleteIf loads the entry from the map if it still exists, then executes the given condition function with the value.
-// If the condition returns true, the entry is deleted from the map, otherwise nothing happens. It is guaranteed that
-// the condition function will only be executed once any in-flight ComputeIfPresent operations for that entry complete.
-// Conversely, once the in-flight operations complete, no new ComputeIfPresent operations will be started for that entry
-// until the condition has been checked. If the entry was deleted, any ComputeIfPresent operations queued for that entry
-// while the condition was being checked will be abandoned. No two executions of DeleteIf can execute in parallel for
-// the same entry.
-func (m *ResourceMap[K, T]) DeleteIf(key K, condition func(key K, value T) bool) {
-	eAny, ok := m.syncMap.Load(key)
-	if !ok {
-		return
-	}
+// ComputeDeletion loads the entry from the map if it still exists, then executes the given condition
+// function with the value. If the condition returns true, the entry is deleted from the map,
+// otherwise nothing happens. As a "compute" function, the condition is executed synchronously, in
+// other words, it is guaranteed that no other "compute" functions are executing on that entry.
+func (m *ResourceMap[K, V]) ComputeDeletion(key K, condition func(value V) (deleteEntry bool)) {
```

Comment:
consider to return a bool about if the entry is deleted?

Reply:
Makes sense, will do